onnxruntime/core/framework/utils.cc

```diff
@@ -209,6 +212,30 @@ const OrtMemoryInfo& FindMemoryInfoForValue(const SessionState& session_state,
 static common::Status CalculateStaticCopyInfoForFeed(const SessionState& session_state,
                                                      const std::string& input_name,
                                                      MLValueCopyInfo& copy_info) {
+#ifdef ENABLE_TRAINING
+  std::vector<SessionState::NodeInfo> node_info_vec;
+  if (session_state.GetInputNodeInfo(input_name, node_info_vec) == Status::OK()) {
+    const auto& node_info = node_info_vec.front();  // all consumers of a feed have the same device so first entry is fine
+
+    if (node_info.p_node == nullptr) {
+      // ignore dummy entry for an input that we didn't find a use of in the graph.
+      return Status::OK();
+    }
+
+    copy_info.target_device = *node_info.device;
+
+  } else {
+    // This input might be for an intermediate tensor for partial graph execution.
+    const auto* exec_plan = session_state.GetExecutionPlan();
```

Comment:
Is this 'might be' or 'is'? If the former, would be good to add something to make sure that it is an intermediate tensor if possible.

Reply:
The first branch checks if its a graph input, the second branch is fall back if it "might be" intermediate tensor, if not we fail at ORT_RETURN_IF_ERROR(name_to_id.GetIdx(input_name, index));

Reply:
Seems vaguer than that though - you fail if it's a completely invalid name. Whether its an intermediate tensor for a partial graph is untested isn't it, so I was wondering whether it could/should do that. guessing it would need a check on whether the value is an output of the YieldOp, but I would also expect that check should happen well above this level of code.

Actually if we have a known set of inputs for partial execution, should they be added to the input node info map in session state? From a cursory look we just lookup that map for device info, and if that's the case the list doesn't need to be limited to the model's inputs.